### DISCUSSION TITLE
c# automatic constructor inheritance

### DISCUSSION THREAD

**Discussion by Floyddotnet** [dotnet/csharplang#4988] [Language Ideas]
Hi, it would be nice if the C# compiler could automatically generate the code for constructor inheritance. this would save a lot of writing.

I would use the following set of rules:

1. if the derived class does not define its own constructor, the compiler generates the same as in the base class.
2. if the derived class defines a constructor, the compiler does not generate any more.

An example:

```
public class Foo
{
   ...
   public Foo() {...}
   public Foo(...) {...}
   ...
}

public class Bar : Foo { }

 // the compiler will add this methods to Bar:

 public Bar(): base() {}
 public Bar(...): base(...) {}
```

ideally you combine this with a language extension to give he implementation an aliases for generics, see:

```
public class Foo : T = ImplementingClass
{
   ...
   public Foo(ILogger<T>) {...}

   // which is translated by the compiler to the following
   public Foo(ILogger<Foo>) {...}
   ...
}

public class Bar : Foo { }

 // the compiler will add this methods to Bar:
 public Bar(ILogger<Bar>) {...}
```

As far as I can see, there is no relevant impact on existing code.

**Comment by svick** [comment]
> As far as I can see, there is no relevant impact on existing code.

I think that changing the meaning of existing code is a significant impact. If something like this was to be added to the language, it would require some indication that you want to inherit non-default constructors.

Also, I think you could get pretty close to what you want today if you wrote [a source generator](https://docs.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/source-generators-overview?WT.mc_id=DOP-MVP-5002735) for this. With that, your code could look something like this:

```c#
public class Foo
{
   ...
   public Foo() {...}
   public Foo(...) {...}
   ...
}

[InheritConstructors]
public partial class Bar : Foo { }

// the source generator will add these constructors to Bar:

public Bar(): base() {}
public Bar(...): base(...) {}
```

**Reply by Floyddotnet** [reply]
> 
> 
> > As far as I can see, there is no relevant impact on existing code.
> 
> I think that changing the meaning of existing code is a significant impact. If something like this was to be added to the language, it would require some indication that you want to inherit non-default constructors.
> 

Why?

1. if the base class has none constructor the compiler will already now generate an empty constructor for you (nothing changed)
2. if the base class has constructors and the devirated class has some too, see rule 2 (no editional constructor will be generated)
3. if the base class has constructor and the devirated class has none, the compiler generates the same as the base class has
 (the only point is that in some, very raw special cases you get trubble if your existing code does not respect that a class has multiple constructors when you use reflection to createn an instance)

> Also, I think you could get pretty close to what you want today if you wrote [a source generator](https://docs.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/source-generators-overview?WT.mc_id=DOP-MVP-5002735) for this. With that, your code could look something like this:
> 
> ```cs
> 
> [InheritConstructors]
> public partial class Bar : Foo { }
> ```

Yes, i can do this (probably) with source generators. This was not the topic. Following the same reasoning, we don't need many extensions for C#, such as default interface methods ( https://docs.microsoft.com/de-de/dotnet/csharp/language-reference/proposals/csharp-8.0/default-interface-methods).
 
if you work a lot with the repository pattern you will quickly get to the point that the constructors of thousands of classes look the same. likewise if you use inheritance a lot.

**Reply by svick** [reply]
> 3\. if the base class has constructor and the [derived] class has none, the compiler generates the same as the base class has

That is the problem. Consider the following [real](https://source.dot.net/#System.Windows.Forms/System/Windows/Forms/ListControl.cs,5651aee9c3fdc326), but significantly simplified code:

```c#
namespace System.Windows.Forms
{
    public class Control
    {
        public Control() { ... }
        public Control(string text) { ... }
        public Control(Control parent, string text) { ... }
        // few other constructors
        // many other members
    }

    public abstract class ListControl : Control
    {
        // no explicit constructor
        // some other members
    }
}
```

If other constructors started to be inherited automatically, then that would change the public API of the `ListControl` class just by recompiling and that is not an acceptable change. And saying that you can easily fix that by making the constructor explicit is not good enough, because it requires everyone to know about this problem and then go through all their code to find if some of their classes have this problem.

> Yes, i can do this (probably) with source generators. This was not the topic.

It is relevant to this topic, because having a good workaround makes a feature less valuable. (Though it doesn't automatically disqualify it from being considered.)

> Following the same reasoning, we don't need many extensions for C#, such as default interface methods.

I don't see how you could implement default interface methods using source generators.

**Reply by Floyddotnet** [reply]
> > 3. if the base class has constructor and the [derived] class has none, the compiler generates the same as the base class has
> 
> That is the problem. Consider the following [real](https://source.dot.net/#System.Windows.Forms/System/Windows/Forms/ListControl.cs,5651aee9c3fdc326), but significantly simplified code:
> 
> ```cs
> namespace System.Windows.Forms
> {
>     public class Control
>     {
>         public Control() { ... }
>         public Control(string text) { ... }
>         public Control(Control parent, string text) { ... }
>         // few other constructors
>         // many other members
>     }
> 
>     public abstract class ListControl : Control
>     {
>         // no explicit constructor
>         // some other members
>     }
> }
> ```
> 
> If other constructors started to be inherited automatically, then that would change the public API of the `ListControl` class just by recompiling and that is not an acceptable change. And saying that you can easily fix that by making the constructor explicit is not good enough, because it requires everyone to know about this problem and then go through all their code to find if some of their classes have this problem.
> 
Why? If the Control needs this additional parameter you have to solve the problem in some way or ListControl won't compile anymore. Otherwise you only get additional constructors for free.

> > Yes, i can do this (probably) with source generators. This was not the topic.
> 
> It is relevant to this topic, because having a good workaround makes a feature less valuable. (Though it doesn't automatically disqualify it from being considered.)

features that are generally useful and have a broad use case should not depend on workarounds. an example: why are there language constructs for loops when i can achieve the same thing with goto? By your definition, they would also be less valuable.

> > Following the same reasoning, we don't need many extensions for C#, such as default interface methods.
> 
> I don't see how you could implement default interface methods using source generators.

Sure, the workaround is to decorate all classes implementing this interface with attributes and rewrite them with source generators too.

**Reply by HaloFour** [reply]
Automatic inheritance of the constructors definitely sounds undesirable to me.  Any existing code that might accidentally result in an automatically inherited constructor is a bug waiting to happen.  IMO, if this language feature were to be considered there should be syntax that the developer can use to explicitly opt-in to developer inheritance.  That is the same tact used by C++:

```cpp
class BaseClass
{
public:
    BaseClass(int value);
};

class DerivedClass : public BaseClass
{
public:
    using BaseClass::BaseClass;  // causes all base constructors to be inherited
};
```

**Reply by Floyddotnet** [reply]
> there should be syntax that the developer can use to explicitly opt-in to developer inheritance. That is the same tact used by C++:
> 
this would at least reduce the boilerplate code

**Reply by CyrusNajmabadi** [reply]
That syntax could then be an attribute that a source generator could use to generate the constructors :-)

**Reply by Floyddotnet** [reply]
> That syntax could then be an attribute that a source generator could use to generate the constructors :-)

source generators were not introduced to make future language extensions obsolete.

**Reply by CyrusNajmabadi** [reply]
They certainly were introduced so that we could have a cheap and effective way to allow problems to be solved without having to rev the language :-)

**Reply by 333fred** [reply]
> source generators were not introduced to make future language extensions obsolete.

No, but they would be an excellent, cheap place to prove out an entirely-additive language change, experiment with different rulesets, and see whether the broader community would be interested in such a change _before_ making an irreversible change to the language.

**Comment by HaloFour** [comment]
See: #806

**Reply by TonyValenti** [reply]
I think that once we get required properties, I'm going to quit using constructors at all.
